rasa/nlu/featurizers/dense_featurizer/lm_featurizer.py

```diff
@@ -1,65 +1,757 @@
 
-    def process(self, message: Message, **kwargs: Any) -> None:
-        """Sets the dense features from the language model doc to the incoming
-        message."""
+        batch_size = 64
+
         for attribute in DENSE_FEATURIZABLE_ATTRIBUTES:
-            self._set_lm_features(message, attribute)
 
-    def _set_lm_features(self, message: Message, attribute: Text = TEXT) -> None:
+            non_empty_examples = list(
+                filter(lambda x: x.get(attribute), training_data.training_examples)
+            )
+
+            batch_start_index = 0
+
+            while batch_start_index < len(non_empty_examples):
+
+                batch_end_index = min(
+                    batch_start_index + batch_size, len(non_empty_examples)
+                )
+                # Collect batch examples
+                batch_messages = non_empty_examples[batch_start_index:batch_end_index]
+
+                # Construct a doc with relevant features
+                # extracted(tokens, dense_features)
+                batch_docs = self._get_docs_for_batch(batch_messages, attribute)
+
+                for index, ex in enumerate(batch_messages):
+                    self._set_lm_features(batch_docs[index], ex, attribute)
+                batch_start_index += batch_size
+
+    def process(self, message: Message, **kwargs: Any) -> None:
+        """Process an incoming message by computing its tokens and dense features.
+
+        Args:
+            message: Incoming message object
+        """
+
+        # process of all featurizers operates only on TEXT and ACTION_TEXT attributes,
+        # because all other attributes are labels which are featurized during training
+        # and their features are stored by the model itself.
+        for attribute in {TEXT, ACTION_TEXT}:
+            if message.get(attribute):
+                self._set_lm_features(
+                    self._get_docs_for_batch(
+                        [message], attribute=attribute, inference_mode=True
+                    )[0],
+                    message,
+                    attribute,
+                )
+
+    def _set_lm_features(self, doc, message: Message, attribute: Text = TEXT) -> None:
         """Adds the precomputed word vectors to the messages features."""
-        doc = self._get_doc(message, attribute)
 
-        if doc is None:
-            return
+        hf_transformers_doc = message.get(LANGUAGE_MODEL_DOCS[attribute])
+        if hf_transformers_doc:
+            # This should only be the case if the deprecated
+            # HFTransformersNLP component is used in the pipeline
+            # TODO: remove this when HFTransformersNLP is removed for good
+            logging.debug(
+                f"{LANGUAGE_MODEL_DOCS[attribute]} set: this "
+                f"indicates you're using the deprecated component "
+                f"HFTransformersNLP, please remove it from your "
+                f"pipeline."
+            )
+            doc = hf_transformers_doc
```

Comment:
Shouldn't this check already happen before line 714 and 732? Otherwise we are constructing the doc all over again, which is double work.

Reply:
If we move it before those lines, we could also just print the warning just once per batch. We could check if the doc is set for the first message in the batch and if yes print the warning. If it is set for one message it should be set for all.